src/firefly/js/visualize/HiPSUtil.js

```diff
@@ -0,0 +1,344 @@
+
+import {get} from 'lodash';
+import {clone} from '../util/WebUtil.js';
+import {CysConverter} from './CsysConverter.js';
+import {makeWorldPt, makeDevicePt} from './Point.js';
+import {SpatialVector, HealpixIndex, Utils} from './projection/aladinProj/HealpixIndex.js';
+import {convert, computeDistance} from './VisUtil.js';
+import {replaceHeader} from './WebPlot.js';
+import {primePlot} from './PlotViewUtil.js';
+
+
+/**
+ *
+ * @param {WebPlot} plot
+ * @param {WorldPt} wp new center of projection
+ */
+export function changeProjectionCenter(plot, wp) {
+    if (!plot) return undefined;
+    wp= convert(wp, plot.projection.coordSys);
+    const header= clone(plot.projection.header, {crval1:wp.x, crval2:wp.y});
+    return replaceHeader(plot,header);
+}
+
+
+
+export function getBestHiPSlevelORGIN(plot) {
+    if (!plot) return null;
+    const screenPix= plot.projection.getPixelScaleArcSec()/plot.zoomFactor;
+
+    if (screenPix> 130) return 'allsky';
+
+    const ratio= screenPix / 51;
+    if (ratio> 1/2) return 4;
+    else if (ratio< 1/2) return 5;
+    else if (ratio< 1/4) return 6;
+    else if (ratio<  1/8) return 6;
+    else if (ratio<  1/16) return 7;
+    else if (ratio<  1/32) return 8;
+    else return 9;
+}
+
+
+
+export function getBestHiPSlevel(plot, limitByMax= false) {
+    if (!plot) return null;
+
+    // const screenPix= plot.projection.getPixelScaleArcSec()/plot.zoomFactor;
+    const {fov}= getPointMaxSide(plot,plot.viewDim);
+
+    const screenPix= (fov/plot.viewDim.width)*3600;
+    if (screenPix> 130) return 'allsky';
+
+    const nside = HealpixIndex.calculateNSide(screenPix*512);
+
+
+    // console.log(fov/ plot.viewDim.width);
+
+    // const nside = HealpixIndex.calculateNSide(screenPix);
+
+    let norder = Math.log(nside)/Math.log(2);
+    norder= Math.max(3, norder);
+
+    if (limitByMax) {
+        const maxOrder= Number(get(plot, 'hipsProperties.hips_order', '3'));
+        norder= Math.min(norder, maxOrder);
+    }
+    return norder;
+
+}
+
+
+export function makeHiPSTileUrl(plot, nOrder, tileNumber) {
+    if (!plot) return null;
+    const dir= Math.floor(tileNumber/10000)*10000;
+    return `${plot.hipsUrlRoot}/Norder${nOrder}/Dir${dir}/Npix${tileNumber}.${getHiPSTileExt(plot)}`;
+}
+
+export function makeHiPSAllSkyUrl(plot) {
+    if (!plot) return null;
+    return `${plot.hipsUrlRoot}/Norder3/Allsky.${getHiPSTileExt(plot)}`;
+}
+
+export function getHiPSTileExt(plot) {
+    const exts= get(plot, 'hipsProperties.hips_tile_format', 'jpg');
+    if (!plot) return null;
+    if (exts.includes('png')) return'png';
+    else if (exts.includes('jpeg') || exts.includes('jpg')) return 'jpg';
+    else return 'jpg';
+
```

Comment:
I have no plans to bring fits over to the client right now.  I think the fits is mostly for reference.  I guess we could do something with the server so that we restretche on the fly.  The point of all this is to use preprocessed data.